#' Simulate Data (Phenotype, Covariates, Genotypes) for Testing Purpose
#'
#' This function takes the desired proportion below LOD (limit of detection), effect size and MAF of the SNP, and sample sizes
#' as the inputs and generates corresponding phenotype, covariate, and genotype files for testing the pipelines.
#' @param cens_prop A non-negative floating-point numeric value indicating the proportion of measurements below LOD.
#' @param betas A vector list of floating-point numeric values indicating the effect size of the SNP, with linear association assumed.
#' @param mafs A vector list of floating-point numeric values less than 0.5 indicating the minor allele frequency of the SNP.
#' @param N An integer indicating the sample size.
#' @param to_truncate Boolean indicating whether to have truncated values (NA) or censored. Just for testing the function merge_data().
#' @importFrom utils write.table
#' @returns Three textfiles - phenotypes.txt, covariates.txt, genotypes.raw - containing the simulated data,
#' typically for testing step2. In genotypes.raw, a dummy RS-ID of the SNP will be created as the SNP ID.
#' @export
simulate_data <- function(cens_prop, betas, mafs, N, to_truncate){

  FIDs <- sample(10000:1000000, size = N, replace = FALSE)
  IIDs <- FIDs # just for testing purpose, place-holder

  genotype_data <- simulate_HWE_geno(FIDs, IIDs, mafs, N)
  write.table(genotype_data, file = "examples.raw", quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t")

  # simulate covariates, say we have only 1 covariate
  covar1 <- 1 + rnorm(N)
  covariate_data <- data.frame(FID = FIDs, IID = IIDs, COVAR1 = covar1)
  write.table(covariate_data, file = "examples.covar", quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t")

  # simulate phenotypes
  phenotypes <- simulate_phenotype(genotypes = genotype_data, betas = betas, covar = covariate_data, cens_prop = cens_prop, to_truncate = to_truncate)
  write.table(phenotypes, file = "examples.pheno", quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t")

  # make a dummy list of the SNPs
  snps <- data.frame(CHR = rep(1, length(mafs)), ID = sub("_.*", "", colnames(genotype_data)[7:ncol(genotype_data)]), A1 = sub(".*_", "", colnames(genotype_data)[7:ncol(genotype_data)]))
  print(colnames(genotype_data))
  write.table(snps, file = "snps.chosen", quote = FALSE, row.names = FALSE, col.names = TRUE, sep = "\t")

}

#' Simulate the Genotypes based on Hardy-Weiberg Equilibrium
#'
#' This function takes the MAF and sample size to generate genotype data for testing.
#' @param FIDs A list of simulated FIDs.
#' @param IIDs A list of simulated IIDs.
#' @param mafs A list floating-point numeric values less than 0.5 indicating the minor allele frequency of the SNPs.
#' @param N An integer indicating the sample size.
#' @importFrom  stats rnorm
#' @returns A dataframe containing the FID, IID, dummy PAT, dummy MAT, dummy SEX, dummy PHENOTYPE, and the SNP with dummy RS-ID.
#' @export
simulate_HWE_geno <- function (FIDs, IIDs, mafs, N){

  PAT <- 2000 + FIDs # just dummy place-holders for matching .raw format, will be deleted in step2
  MAT <- 5000 + FIDs
  SEX <- sample(c(0,1), size = N, replace = TRUE)
  PHENOTYPE <- 1 # just dummy place-holder, won't be the same as in simulate_phenotype()

  genotypes <- data.frame(FID = FIDs, IID = IIDs, PAT = PAT, MAT = MAT, SEX = SEX, PHENOTYPE = PHENOTYPE)

  for (maf in mafs){
    q <- maf
    p <- 1-q

    GT <- as.numeric(sample(c("0","1","2"),size=N,replace=TRUE,prob=c(p^2,2*p*q,q^2))) # will have the dummy column ID

    genotypes <- cbind(genotypes, GT)
    colnames(genotypes)[ncol(genotypes)] <- paste0("rs", sample(100:10000, size = 1), "_", sample(c("A", "T", "G", "C"), size = 1))

  }

  return(genotypes)
}

#' Simulate Phenotype using the Genotypes and their Corresponding Effect Sizes
#'
#' @param genotypes The dataframe generated by simulate_HWE_geno().
#' @param betas The effect sizes of each SNP
#' @param covar The covariate, in the simulation here we have only 1 covariate for simplicity
#' @param cens_prop The proportion of measurements that should be below the LOD.
#' @param to_truncate Boolean indicating whether to have truncated values (NA) or censored. Just for testing the function merge_data().
#' @importFrom stats rnorm qnorm sd
#' @returns A dataframe containing the FID, IID, censored status, and measured values.
#' @export
simulate_phenotype <- function(genotypes, betas, covar, cens_prop, to_truncate){
  genos <- genotypes[,7:ncol(genotypes)] # in the .raw-formatted dataframe, SNPs start from the 7th column
  X <- as.matrix(cbind(1, genos)) # design matrix
  b <- c(1, betas) # coefficients, vector
  h <- as.numeric(X %*% b) # linear predictor
  covar <- covar[,3]
  y_norm <- h + covar * 0.05 + rnorm(nrow(X)) #here let's assume the effect size of covariate1 is 0.05
  y_norm <- y_norm - min(y_norm) # make positive, not really necessary though...

  # add LOD
  lod <- qnorm(cens_prop, mean = mean(y_norm), sd = sd(y_norm))
  cens_status <- rep(1, nrow(X))
  cens_status[y_norm < lod] <- 0
  measurements <- y_norm
  if (to_truncate){
    measurements[measurements<lod] <- NA
  }else{
    measurements[measurements<lod] <- lod
  }

  phenotypes <- data.frame(FID = genotypes$FID, IID = genotypes$IID, cens_status = cens_status, measurements = measurements)
  return(phenotypes)
}
