#' Run Linear Model with PLINK2
#'
#' This function runs the linear model GWAS, as Step 1, using PLINK2 with the unextracted genotype files and the prepared covariate and phenotype files
#' and outputs the GWAS outcomes
#' @param pheno_prepared REQUIRED. The path to the phenotype file, prepared by merge_data() which has been filtered for the selected IIDs.
#' @param covar_prepared REQUIRED. The path to the covariate file, prepared by merge_data() which has been filtered for the selected IIDs.
#' @param file_calls REQUIRED. A string representing the genotype call file, ending with .pgen or .bed.
#' @param file_variant_info REQUIRED. A string representing the variant information files, ending with .pvar or .bim.
#' @param file_sample_info REQUIRED. A string representing the sample information file, ending with .psam or .fam.
#' @param thres_hwe OPTIONAL. Threshold for Hardy-Weignberg equilibrium, default is 1e-20.
#' @param thres_maf OPTIONAL. Threshold for MAF of the analyzed SNPs, default is 0.01.
#' @param thres_geno OPTIONAL. Threshold of missing call rates for filtering out variants. Default is 0.05.
#' @param ci OPTIONAL. Confidence intervals with the given width to be reported for each beta or odds-ratio. Default is 0.95.
#' @param plink_file_version REQUIRED. Version of plink files, default is 2
#' @param output_name OPTIONAL. A string of the name for the output file, can be the group name or chromosome number.
#' Default is "out". The suffix "glm.linear" will be added automatically by PLINK.
#' @returns GWAS outcomes as text files, generated by PLINK2.
step1_plink <- function(pheno_prepared = NULL, covar_prepared = NULL, file_calls = NULL, file_variant_info = NULL, file_sample_info = NULL, thres_hwe = 1e-20, thres_maf = 0.01, thres_geno = 0.05, ci = 0.95, plink_file_version, output_name = "out"){

  message("Trying to run Step 1 Linear Model GWAS with PLINK2...")

  if (output_name != "out"){
    output_dir <- dirname(output_name)
    if (output_dir=="."){
      output_dir <- getwd()
    }
  }else{
    output_dir <- getwd()
  }

  if (plink_file_version == 2){
    plink_cmd <- paste0("plink2 --glm hide-covar cols=+a1freq,+ax,+beta --pgen ", file_calls, " --psam ", file_sample_info, " --pvar ", file_variant_info, " --pheno ", pheno_prepared, " --pheno-name tf_lin --covar ", covar_prepared, " --hwe ", thres_hwe, " --maf ", thres_maf, " --geno ", thres_geno, " --ci ", ci, " --out ", output_name)
  }else{
    plink_cmd <- paste0("plink2 --glm hide-covar cols=+a1freq,+ax,+beta --bed ", file_calls, " --fam ", file_sample_info, " --bim ", file_variant_info, " --pheno ", pheno_prepared, " --pheno-name tf_lin --covar ", covar_prepared, " --hwe ", thres_hwe, " --maf ", thres_maf, " --geno ", thres_geno, " --ci ", ci, " --out ", output_name)
  }
  exit_status <- system(plink_cmd)
  if (exit_status !=0){
    warning("Error: PLINK2 encountered an error and exited with status ", exit_status, ". Step 1 GWAS with linear model failed.")
    return(NULL)
  }
  else{
    message("Completed Step 1 GWAS with linear model. Results saved to ", output_dir, ".")
    return(0)
  }
}

#' Clump and organize GWAS Outcomes
#'
#' This function takes the GWAS results (generate by step1_plink() or provided by user) and performs clumping with PLINK2
#' outputs the dataframe containing the clumped SNPs with chr no., position, SNP ID, effect alleles, effect size, standard error, p-value
#' @param gwas_results REQUIRED. A string representing the file containing GWAS results.
#' @param p1 OPTIONAL. The index variant p-value threshold. Default is 5e-8.
#' @param p2 OPTIONAL. The SP2 column p-value threshold. Default is 1e-4.
#' @param r2 OPTIONAL. The r^2 (correlation between genetic variants to reduce redundancy) threshold. Default is 0.1.
#' @param kb OPTIONAL. The maximum distance from the lead variant for SNPs to be considered to be clumped with. Default is 1000.
#' @param file_calls REQUIRED. A string representing the genotype call file, ending with .pgen or .bed.
#' @param file_variant_info REQUIRED. A string representing the variant information files, ending with .pvar or .bim.
#' @param file_sample_info REQUIRED. A string representing the sample information file, ending with .psam or .fam.
#' @param plink_file_version REQUIRED. Version of PLINK files. Integer 1 or 2. Default is 2.
#' @param output_name OPTIONAL. A string of the name for the output file, can be the group name or chromosome number.
#' Default is "out". The suffix ".clumps" will be added automatically by PLINK.
#' @param snp_id_field OPTIONAL. The field (column) name in the GWAS result files that contains the SNP IDs, default is 'ID'.
#' @param p_field OPTIONAL. The field (column) name in the GWAS result files that contains the p-values, default is 'P'.
#' @param a1_field OPTIONAL. The field (column) name in the GWAS result files that contain the effect allele. Default is 'A1'.
#' @param a1_freq_field OPTIONAL. The field (column) name in the GWAS result files that contain the A1 allele frequency. Defeult is 'A1_FREQ'.
#' @param beta_field OPTIONAL. The field (column) name in the GWAS result files that contain the effect estimate (slope). Default is 'BETA'.
#' @param se_field OPTIONAL. The field (column) name in the GWAS result files that contain the standard error. Default is 'SE'.
#' @param pos_field OPTIONAL. The field (column) name in the GWAS result files that contain the SNP position. Default is 'POS'.
#' @importFrom utils read.table
#' @returns The clumped files (ending with .clumps) and the dataframe containing the clumped SNPs (Chr No., Position, SNP ID, effect allele A1, effect size BETA, standard error, and p-value); or, NULL if process failed.
clump_gwas <- function(gwas_results = NULL, p1 = 5e-8, p2 = 1e-4, r2 = 0.1, kb = 1000, file_calls = NULL, file_variant_info = NULL, file_sample_info = NULL, plink_file_version,  output_name = "out",
                       snp_id_field = "ID", p_field = "P", a1_field = "A1", a1_freq_field = "A1_FREQ", beta_field = "BETA", se_field = "SE", pos_field = "POS"){


  # get output directory just for the output message
  if (output_name != "out"){
    output_dir <- dirname(output_name)
    if (output_dir=="."){
      output_dir <- getwd()
    }
  }else{
    output_dir <- getwd()
  }

  if (plink_file_version == 2){
    plink_cmd <- paste0("plink2 --clump ",  gwas_results, " --pgen ", file_calls, " --psam ", file_sample_info, " --pvar ", file_variant_info, " --clump-r2 ", r2, " --clump-kb ", kb, " --clump-p1 ", p1, " --clump-p2 ", p2, " --clump-id-field ", snp_id_field, " --clump-p-field ", p_field, " --clump-a1-field ", a1_field, " --out ", output_name)
  }else{
    plink_cmd <- paste0("plink2 --clump ",  gwas_results, " --bed ", file_calls, " --fam ", file_sample_info, " --bim ", file_variant_info, " --clump-r2 ", r2, " --clump-kb ", kb, " --clump-p1 ", p1, " --clump-p2 ", p2, " --clump-id-field ", snp_id_field, " --clump-p-field ", p_field, " --clump-a1-field ", a1_field, " --out ", output_name)
  }
  exit_status <- system(plink_cmd)
  if (exit_status !=0){
    warning("Error: PLINK2 encountered an error and exited with status ", exit_status, ". Clumping failed.")
    return(NULL)
  }else{

    # PLINK give .clumps and .log, we want specifically the one with .clumps
    base_name <- basename(output_name)
    matched_clump <- list.files(path = output_dir, pattern = paste0(base_name, ".*\\.clumps"), full.names = TRUE)
    if (length(matched_clump) != 0){
      message("Successfully clumped. Results saved in ", output_dir, ".")

      # organize the clumped results into a dataframe,
      # by finding the clumped results in the output directory, reading the SNP IDs,
      # to match them back in the GWAS output to get the original estimates as well as the variant information

      # load the clumped file
      clumped <- read.table(matched_clump, header = TRUE, comment.char="")
      # load the GWAS result
      original <- read.table(gwas_results, header = TRUE, comment.char="")
      # get the list of snps
      clumped_list <- clumped[, 3]
      # shrink the GWAS results
      filtered_original <- subset(original, original[[snp_id_field]] %in% clumped_list)
      rm(clumped, original)
      # the dataframe storing the clumped results; for the gwas result, the first column is always the chromosome number 'CHROM'
      clumped_snps <- data.frame(CHR=filtered_original[,1], POS=filtered_original[,c(pos_field)],
                                 ID=filtered_original[, c(snp_id_field)], A1=filtered_original[,c(a1_field)], A1_FREQ=filtered_original[,c(a1_freq_field)],
                                 BETA_LINEAR=filtered_original[,c(beta_field)], SE_LINEAR = filtered_original[, c(se_field)], P_LINEAR=filtered_original[,c(p_field)])


    }else{
      # if nothing shows up in the clumped results, there will not be an output by PLINK
      warning("Successfully clumped. But no SNPs matched the clumping criteria.")
      return(NULL)
    }

    return(clumped_snps)

  }

}

#' List SNPs for Extraction from the Clumped Results
#'
#' This function gets the Chromosome Number, SNP ID, and the effect allele A1 from the clumped dataframe generated by organize_clumped() or
#' and generates the dataframe containing the three columns described for get_geno()
#' @param clumped_df REQUIRED. The dataframe of the clumped results, generated by clump_gwas()
#' @returns dataframe containing chromosome number, SNP ID, A1
snps_to_extract <- function(clumped_df){
  snp_df <- data.frame(CHR = clumped_df$CHR, ID = clumped_df$ID, A1 = clumped_df$A1)
}
