#' Refine Estimates with Tobit Model
#'
#' This function takes the converted genotype, a dataframe of covariate and phenotype,
#' and the table of snps selected either as the clumping results or three-column table containing only the chromosome number, SNP ID and effect allele
#' and gives the refined effect estimates (intercept, BETA and SE, statistic, p-value) with Tobit model in textfile,
#' @param geno_converted REQUIRED. A string representing the converted genotype file (.raw) generated by get_geno().
#' @param pheno_covar REQUIRED. The datafame containing the FID, IID, phenotypes and covariates, organized by merge_data().
#' @param snps_to_refine REQUIRED. The dataframe containing the clumped results or selected snps for refinement, clumping generated by either step1_plink(), or SNPs provided by users themselves.
#' @param output_name OPTIONAL. A string for the output file (.txt) name containing the refined result. Default is "out"
#' @importFrom utils write.table read.table
#' @importFrom stats as.formula
#' @importFrom censReg censReg
#' @importFrom broom tidy
#' @returns A text file including the refined estimates of the SNPs using Tobit model.
#' @export
step2_run_tobit <- function(geno_converted = NULL, pheno_covar = NULL, snps_to_refine = NULL, output_name = "out"){

  if (output_name != "out"){
    output_dir <- dirname(output_name)
    if (output_dir=="."){
      output_dir <- getwd()
    }
  }else{
    output_dir <- getwd()
  }

  # read the genotype data
  geno_data <- read.table(geno_converted, header =TRUE)
  # from the 7th header they start to be the SNP IDs, but .raw file has them as rs..._A1
  # we only want the part before _A1
  geno_data <- geno_data[, -c(1, 3:6)] # exclude the FID as well as columns that are not the SNPs
  colnames(geno_data)[2:ncol(geno_data)] <- sub("_.*", "", colnames(geno_data)[2:ncol(geno_data)])

  # merge the phenotype-covariate data with the genotype data; pheno_covar has complete cases
  message(nrow(pheno_covar), " individuals provided in phenotype/covariate dataset.")
  # column 1 and 2 are FID and IID, 3 the censored status (binary), 4 the censored measurements, 5 for Tobit, 6 for linear
  if (colnames(pheno_covar)[6] == "tf_lin"){
    pheno_covar <- pheno_covar[,-6] # remove the transformed values for linear for consistency in subsequent runs
  }

  merged_data <- merge(pheno_covar, geno_data, by = "IID") # order the participants by merging the datasets

  # list out the IDs for testing
  test_snps <- snps_to_refine$ID
  num_snps <- length(test_snps)

  message(num_snps, " SNP(s) to refine in total.")

  # covariate part of the formula, directly pasted by collapsing with "+"
  covar_columns <- colnames(merged_data)[6:(ncol(merged_data)-num_snps)]
  covar_parts <- paste(covar_columns, collapse = " + ")

  # dataframe to store the regression results
  df_reg <- data.frame(matrix(nrow = num_snps, ncol = 6)) # ID, intercept, beta, se, stat, p
  colnames(df_reg) <- c("ID", "INTERCEPT", "BETA_TB","SE_TB", "STAT", "P_TB")
  df_reg$ID <- test_snps
  for (i in 1:num_snps){
    snp <- test_snps[i]
    message("Testing SNP no.", i, ": ", snp, "...")

    # edit the regression formula each snp
    # tf_tb is the column name of the transformed censored measurements,
    eq <- as.formula(paste("tf_tb ~ ", snp, "+", covar_parts)) # put the snp first so that the 2nd row of the model result is the estimates for the snp
    lod_transformed <- min(merged_data$tf_tb)

    tb_model <- censReg(eq, data = merged_data, left = lod_transformed)

    tb_organized <- data.frame(tidy(tb_model))

    # put the estimates of the SNP into the dataframe df_reg
    df_reg[i, 2] <- tb_organized[1, "estimate"]
    df_reg[i, 3] <- tb_organized[2, "estimate"]
    df_reg[i, 4] <- tb_organized[2, "std.error"]
    df_reg[i, 5] <- tb_organized[2, "statistic"]
    df_reg[i, 6] <- tb_organized[2, "p.value"]
  }

  snps_to_refine <- merge(snps_to_refine, df_reg, by = "ID")
  message("Finised refining the estimates, result writing to ", output_dir)

  write.table(snps_to_refine, file = paste0(output_name, ".refined"), row.names = FALSE, col.names=TRUE, quote = FALSE, sep = "\t")

}
